src/V3Randomize.cpp

```diff
@@ -0,0 +1,280 @@
+
+    BaseToDerivedMap m_baseToDerivedMap;  // Mapping from base classes to classes that extend them
+
+    // METHODS
+    VL_DEBUG_FUNC;
+
+    void markMembers(AstClass* nodep) {
+        for (auto* classp = nodep; classp;
+             classp = classp->extendsp() ? classp->extendsp()->classp() : nullptr) {
+            for (auto* memberp = classp->stmtsp(); memberp; memberp = memberp->nextp()) {
+                // If member is rand and of class type, mark its class
+                if (VN_IS(memberp, Var) && VN_CAST(memberp, Var)->isRand()) {
+                    if (auto* classRefp = VN_CAST(memberp->dtypep(), ClassRefDType)) {
+                        auto* classp = classRefp->classp();
+                        markMembers(classp);
+                        markDerived(classp);
+                        classRefp->classp()->user1(true);
+                    }
+                }
+            }
+        }
+    }
+    void markDerived(AstClass* nodep) {
+        if (m_baseToDerivedMap.find(nodep) != m_baseToDerivedMap.end()) {
+            for (auto* classp : m_baseToDerivedMap.at(nodep)) {
+                classp->user1(true);
+                markMembers(classp);
+                markDerived(classp);
+            }
+        }
+    }
+    void markAllDerived() {
+        for (auto p : m_baseToDerivedMap) {
+            if (p.first->user1()) markDerived(p.first);
+        }
+    }
+
+    // VISITORS
+    virtual void visit(AstClass* nodep) override {
+        iterateChildren(nodep);
+        if (nodep->extendsp()) {
+            // Save pointer to derived class
+            auto* basep = nodep->extendsp()->classp();
+            m_baseToDerivedMap[basep].insert(nodep);
+        }
+    }
+    virtual void visit(AstMethodCall* nodep) override {
+        iterateChildren(nodep);
+        if (nodep->name() != "randomize") return;
+        if (AstClassRefDType* classRefp = VN_CAST(nodep->fromp()->dtypep(), ClassRefDType)) {
+            auto* classp = classRefp->classp();
+            classp->user1(true);
+            markMembers(classp);
+        }
+    }
+    virtual void visit(AstNode* nodep) override { iterateChildren(nodep); }
+
+public:
+    // CONSTRUCTORS
+    explicit RandomizeMarkVisitor(AstNetlist* nodep) {
+        iterate(nodep);
+        markAllDerived();
+    }
+    virtual ~RandomizeMarkVisitor() override = default;
+};
+
+//######################################################################
+// Visitor that defines a randomize method where needed
+
+class RandomizeVisitor final : public AstNVisitor {
+private:
+    // NODE STATE
+    // Cleared on Netlist
+    //  AstClass::user1()       -> bool.  Set true to indicate needs randomize processing
+    // AstUser1InUse    m_inuser1;      (Allocated for use in RandomizeMarkVisitor)
+
+    // TYPES
+    typedef std::unordered_map<AstNodeDType*, AstVar*> ValueTableMap;
+
+    // STATE
+    ValueTableMap m_enumValueTabMap;  // Tables with enum values
```

Comment:
Suspect this can be a AstNodeDType::user2p() instead.  Remember to update NODE STATE comment and add AstUser2InUse if so.